crypto/cipher_extra/aead_test.cc

```diff
@@ -1364,6 +1364,48 @@ TEST(AEADTest, TestGCMSIV256Change16Alignment) {
   free(encrypt_ctx_256);
 }
 
+TEST(AEADTest, TestMonotonicityCheck) {
+
+  static const uint8_t kEvpAeadCtxKey[32] = {0};
+
+  // Only the tls13() ciphers have monotonicity checks
+  struct {
+    const EVP_AEAD *cipher;
+    const size_t key_len;
+  } ctx[] = { { EVP_aead_aes_128_gcm_tls13(), 16},
+              { EVP_aead_aes_256_gcm_tls13(), 32} };
+
+  for (int i = 0; i < 2; i++) {
+    const EVP_AEAD *cipher = ctx[i].cipher;
+    EVP_AEAD_CTX *encrypt_ctx =
+      (EVP_AEAD_CTX *)malloc(sizeof(EVP_AEAD_CTX) + 8);
+    ASSERT_TRUE(encrypt_ctx);
+
+    EVP_AEAD_CTX_zero(encrypt_ctx);
+    ASSERT_TRUE(EVP_AEAD_CTX_init(encrypt_ctx, cipher, kEvpAeadCtxKey, ctx[i].key_len, 16, NULL))
+        << ERR_error_string(ERR_get_error(), NULL);
+
+    uint8_t nonce[12] = {0};
+    uint8_t last_byte = sizeof(nonce) - 1;
+    uint8_t plaintext[16] = {0};
+    uint8_t ciphertext[32] = {0};
+    size_t out_len = 0;
+
+    // Checks that sequence numbers are allowed to increment by more than one
+    // as long as monotonicity is preserved. Here the implicit IV is presumed
+    // to be a zero-filled array. That lets us update the nonce value directly
+    // with an increasing sequence number.
+    for (size_t sequence_num = 0; sequence_num <= 255; sequence_num+=10) {
+      nonce[last_byte] = sequence_num;
```

Comment:
NP: this should probably use uint8_t, currently it looks a little weird to assign a size_t to a byte array.

Reply:
🤔 Then I end up with this compiler error.
```
/awslc/crypto/cipher_extra/aead_test.cc:1390:49: error: comparison is always true due to limited range of data type [-Werror=type-limits]
 1390 |     for (uint8_t sequence_num = 0; sequence_num <= 255; sequence_num+=10) {
      |                                    ~~~~~~~~~~~~~^~~~~~
```

Reply:
Ahhh yeah didn't think about that, this works then.